### PR TITLE
IDP-499 - Otel tracing

### PR DESCRIPTION
**Pull request by Gcaya**
## Context
We want to have open telemetry built into the library so people can have easy to understand traces. We also want to link traces from producer and consumer.

## What changed
- Added behavior for publisher that starts a producer activity and applies the distributed extensions on every event
- Modified EventPropagationClient to use behavior
- Added behavior for subscriber that retrieves the distributed extensions and links them to the consumer activity
- Modified EventGridWebhookHandler to use behavior
- Added tests to make sure the behavior were being used
- 

## What's to be done
- Integration AppInsight alongside otel
- Add more tests on behaviors

### CODE REVIEW

File: src/Workleap.DomainEventPropagation.Tests/InMemoryActivityTracker.cs
Comment (decarufe) [inline src/Workleap.DomainEventPropagation.Tests/InMemoryActivityTracker.cs:1]: Should you follow [lock guidelines](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/lock#guidelines)?

**Review by decarufe — APPROVED** [review]

File: src/Workleap.DomainEventPropagation.Tests/Subscription/IntegrationTests.cs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+                ["AzureAd:TenantId"] = "50103d8b-4c52-4c40-9f78-1461000290d7",
+                ["AzureAd:ClientId"] = "cfde6263-aa07-4ae0-a054-66f536cfdc7b",
+                ["AzureAd:Audience"] = "cfde6263-aa07-4ae0-a054-66f536cfdc7b",
+            });
+
+        // Here begins services registrations in the dependency injection container
+        webAppBuilder.Services.AddLogging(x =>
+            x.SetMinimumLevel(LogLevel.Debug)
+                .ClearProviders());
+                // .AddProvider(new XunitLoggerProvider(this._testOutputHelper)));
```
[the remark this answers is not in the document]
Comment (Gcaya) [reply src/Workleap.DomainEventPropagation.Tests/Subscription/IntegrationTests.cs:69]: Woops, this file was not supposed to make it here 😅

File: src/Workleap.DomainEventPropagation.Publishing/EventPropagationClient.cs [hunk 2]
```diff
@@ -16,13 +17,16 @@ internal sealed class EventPropagationClient : IEventPropagationClient
 
     private readonly EventPropagationPublisherOptions _eventPropagationPublisherOptions;
     private readonly IAzureClientFactory<EventGridPublisherClient> _eventGridPublisherClientFactory;
+    private readonly IServiceProvider _serviceProvider;
 
     public EventPropagationClient(
         IAzureClientFactory<EventGridPublisherClient> eventGridPublisherClientFactory,
-        IOptions<EventPropagationPublisherOptions> eventPropagationPublisherOptions)
+        IOptions<EventPropagationPublisherOptions> eventPropagationPublisherOptions,
+        IServiceProvider serviceProvider)
```
[the remark this answers is not in the document]
Comment (Gcaya) [reply src/Workleap.DomainEventPropagation.Publishing/EventPropagationClient.cs:25]: ah indeed, I could change it for the IEnumerable of services I need

File: src/Workleap.DomainEventPropagation.Tests/InMemoryActivityTracker.cs
Comment (Gcaya) [reply src/Workleap.DomainEventPropagation.Tests/InMemoryActivityTracker.cs:1]: Yeah I can change that!

**Review by decarufe — APPROVED** [review]

**Pull request merged**